File: dataforge-core/src/main/java/hep/dataforge/data/DataFilter.java
package hep.dataforge.data;

/**
 * A filter that could select a subset from a DataNode without changing its type.
 */
public interface DataFilter {
    DataFilter IDENTITY = new DataFilter() {
        @Override
        public <T> DataNode<T> filter(DataNode<T> node) {
            return node;
        }
    };

    static DataFilter byPattern(String pattern) {
        return new DataFilter() {
            @Override
            public <T> DataNode<T> filter(DataNode<T> node) {
                DataSet.Builder<T> builder = DataSet.builder(node.type());
                builder.setName(node.getName());
                node.dataStream(true)
                        .filter(d -> d.getName().matches(pattern))
                        .forEach(builder::putData);
                return builder.build();
            }
        };
    }

    /**
     * Perform a selection. Resulting node contains references to the data in the initial node.
     * Node structure and meta is maintained if possible.
     *
     * @param node
     * @param <T>
     * @return
     */
    <T> DataNode<T> filter(DataNode<T> node);
}

File: dataforge-core/src/main/kotlin/hep/dataforge/data/DataFilter.kt
package hep.dataforge.data

/**
 * A filter that could select a subset from a DataNode without changing its type.
 */
interface DataFilter {

    /**
     * Perform a selection. Resulting node contains references to the data in the initial node.
     * Node structure and meta is maintained if possible.
     *
     * @param node
     * @param <T>
     * @return
     */
    fun <T: Any> filter(node: DataNode<T>): DataNode<T>

    companion object {
        val IDENTITY: DataFilter = object : DataFilter {
            override fun <T: Any> filter(node: DataNode<T>): DataNode<T> {
                return node
            }
        }

        fun byPattern(pattern: String): DataFilter {
            return object : DataFilter {
                override fun <T: Any> filter(node: DataNode<T>): DataNode<T> {
                    return DataSet.edit(node.type).apply {
                        name = node.name
                        node.dataStream(true)
                                .filter { d -> d.name.matches(pattern.toRegex()) }
                                .forEach { add(it) }

                    }.build()
                }
            }
        }
    }
}
